src/Time/TimeSteppers/AdamsBashforth.cpp

```diff
@@ -374,6 +374,148 @@ void AdamsBashforth::boundary_dense_output_impl(
   return boundary_impl(result, coupling, ApproximateTime{time});
 }
 
+namespace {
+template <typename T>
+class SmallStepIterator {
+ public:
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = Time;
+  using pointer = const Time*;
+  using reference = const Time&;
+  using difference_type = std::ptrdiff_t;
+
+  enum class Side { Local, Remote, Both };
+
+  SmallStepIterator() = default;
+
+  SmallStepIterator(const bool time_runs_forward,
+                    typename BoundaryHistoryEvaluator<T>::iterator local_begin,
+                    typename BoundaryHistoryEvaluator<T>::iterator remote_begin,
+                    typename BoundaryHistoryEvaluator<T>::iterator local_end,
+                    typename BoundaryHistoryEvaluator<T>::iterator remote_end)
+      : before_{time_runs_forward},
+        local_time_(std::move(local_begin)),
+        remote_time_(std::move(remote_begin)),
+        local_end_(std::move(local_end)),
+        remote_end_(std::move(remote_end)) {}
+
+  reference operator*() const {
+    return std::max(*local_time_, *remote_time_, before_);
+  }
+  pointer operator->() const { return &**this; }
+
+  Side side() const {
+    if (before_(*local_time_, *remote_time_)) {
+      return Side::Remote;
+    } else if (before_(*remote_time_, *local_time_)) {
+      return Side::Local;
+    } else {
+      return Side::Both;
+    }
+  }
+
+  // These are the m^s(n) in the paper.
+  const typename BoundaryHistoryEvaluator<T>::iterator& local_iterator() const {
+    return local_time_;
+  }
+  const typename BoundaryHistoryEvaluator<T>::iterator& remote_iterator()
+      const {
+    return remote_time_;
+  }
+
+  SmallStepIterator& operator++() {
+    ASSERT(local_time_ != local_end_ and remote_time_ != remote_end_,
+           "Overran iterator");
+    auto local_candidate = std::next(local_time_);
+    auto remote_candidate = std::next(remote_time_);
+
+    if (local_candidate == local_end_ and remote_candidate == remote_end_) {
```

Comment:
Clang-tidy is complaining that there are multiple branches with the same code in each branch. It wants you to combine them (untested)

```cpp
if (local_candidate == local_end_ and remote_candidate == remote_end_) {
  local_time_ = std::move(local_candidate);
  remote_time_ = std::move(remote_candidate);
} else if (local_candidate == local_end_ or before_(*remote_candidate, *local_candidate)) {
  remote_time_ = std::move(remote_candidate);
} else if (remote_candidate == remote_end_ or before_(*local_candidate, *remote_candidate)) {
  local_time_ = std::move(local_candidate);
} else {
  local_time_ = std::move(local_candidate);
  remote_time_ = std::move(remote_candidate);
}
```

However, this might still complain because the first and last branch are identical still...

Reply:
You've introduced a bug by reordering the cases.  If `local_candidate != local_end_` and `remote_candidate == remote_end_` you're likely hitting UB.

Reply:
Not surprised as I have no clue what's going on here. If you're unable to find a way to combine branches I'd say just `NOLINT` this whole section

Reply:
All the variables are iterators.  (OK, technically not since `before_` is a variable, too.)

Reply:
Yea, I agree that it looks like clang-tidy wants you to create bugs in your code 🙃 I think what it's worried about is that the branches have the same content and need to be kept in sync. I think reasonable ways to go about silencing it:
1. `NOLINT`
2. a lambda for shared code (though seems a bit silly to have a 3 line lambda definition for a 1 line move)
3. you might be able to actually combine 437 & 439 without breaking logic

Reply:
1. That's what I'm going with, assuming I can figure out which lines to mark.
2. I assume it would then complain that I'm calling the same lambda in two branches.
3. I could.  If it solved the whole problem I might have done that, but since there are two other things to NOLINT I'm going to leave it because I think it's more readable as-is.

Reply:
Sounds good! :)